runtime/ops/tls.rs

```diff
@@ -70,6 +88,457 @@ impl StoresClientSessions for ClientSessionMemoryCache {
+          Ok(0) if self.tls.is_handshaking() => {
+            let err =
+              io::Error::new(ErrorKind::UnexpectedEof, "tls handshake eof");
+            return Poll::Ready(Err(err));
+          }
+          Ok(0) => self.rd_shut = Shut::TcpShut,
+          Ok(_) if matches!(self.rd_shut, Shut::TlsShut | Shut::TcpShut) => {
+            // Just like TCP, when data arrives after the connection has been
+            // closed it means that the connection wasn't closed in a clean way.
+            let err = io::Error::from(ErrorKind::ConnectionReset);
+            return Poll::Ready(Err(err));
+          }
+          Ok(_) => match self.tls.process_new_packets() {
+            // Do a zero-length plaintext read so we can detect the arrival of
+            // 'CloseNotify' messages, even if only the write half is active.
+            // Actually reading data from the socket is done in `poll_read()`.
+            Ok(()) => match self.tls.read(&mut []) {
+              Ok(0) => {}
+              Ok(_) => unreachable!(),
+              // Rustls `read()` returns `ConnectionAborted` when it receives a
+              // 'CloseNotify' alert; this indicates that the remote end of the
+              // connection has initiated a graceful shutdown at the TLS level.
+              Err(err) if err.kind() == ErrorKind::ConnectionAborted => {
+                self.rd_shut = Shut::TlsShut;
+              }
+              Err(err) => return Poll::Ready(Err(err)),
+            },
+            Err(err) => {
+              let err = io::Error::new(ErrorKind::InvalidData, err);
+              return Poll::Ready(Err(err));
+            }
+          },
+          Err(err) if err.kind() == ErrorKind::WouldBlock => {
+            match self.tcp.poll_read_ready(cx) {
+              Poll::Ready(result) => result?,
+              Poll::Pending => tcp_read_ready = false,
+            }
+          }
+          Err(err) => return Poll::Ready(Err(err)),
+        }
+      }
+
+      if tcp_write_ready && self.tls.wants_write() {
+        continue;
+      }
+
+      let poll_again = match flow {
+        _ if self.tls.is_handshaking() => true,
+        Flow::Read => self.tls.wants_read() && self.rd_shut != Shut::TcpShut,
+        Flow::Write => self.tls.wants_write(),
+      };
+      return match poll_again {
+        false => Poll::Ready(Ok(())),
+        true => Poll::Pending,
+      };
+    }
+  }
+
+  fn poll_graceful_close(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+    let p1 = self.poll_io(cx, Flow::Read);
+    let p2 = Pin::new(self).as_mut().poll_shutdown(cx);
+    let _ = ready!(p1);
+    let _ = ready!(p2);
+    Poll::Ready(())
+  }
+}
+
+impl AsyncRead for TlsStream {
+  fn poll_read(
+    mut self: Pin<&mut Self>,
+    cx: &mut Context<'_>,
+    buf: &mut ReadBuf<'_>,
+  ) -> Poll<io::Result<()>> {
+    ready!(self.poll_io(cx, Flow::Read))?;
+
+    if matches!(self.rd_shut, Shut::TlsShut | Shut::TcpShut) {
```

Comment:
I think it obscures the shutdown status of tcp stream and tls stream. When eof occurs in the tcp stream, there may still be readable data in the buffer of the tls stream.